src/corehost/cli/ijwhost/ijwthunk.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "ijwhost.h"
+#include "bootstrap_thunk_chunk.h"
+#include "corhdr.h"
+#include "error_codes.h"
+#include "trace.h"
+#include <heapapi.h>
+#include <new>
+#include <mutex>
+
+#ifdef _WIN64
+#define COR_VTABLE_PTRSIZED     COR_VTABLE_64BIT
+#define COR_VTABLE_NOT_PTRSIZED COR_VTABLE_32BIT
+#else
+#define COR_VTABLE_PTRSIZED     COR_VTABLE_32BIT
+#define COR_VTABLE_NOT_PTRSIZED COR_VTABLE_64BIT
+#endif
+
+namespace
+{
+    std::mutex g_thunkChunkLock{};
+    bootstrap_thunk_chunk* g_pVtableBootstrapThunkChunkList = nullptr;
+}
+
+HANDLE g_heapHandle;
+
+bool patch_vtable_entries(PEDecoder& pe)
+{
+    if (pe.IsILOnly())
+    {
+        // Nothing to do if the PE is IL-only.
+        return true;
+    }
+    
+    size_t numFixupRecords;
+    IMAGE_COR_VTABLEFIXUP* pFixupTable = pe.GetVTableFixups(&numFixupRecords);
+
+    if (numFixupRecords == 0)
+    {
+        // If we have no fixups, no need to allocate thunks.
+        return true;
+    }
+
+    size_t numThunks = 0;
+    for (size_t i = 0; i < numFixupRecords; ++i)
+    {
+        numThunks += pFixupTable[i].Count;
+    }
+
+    size_t chunkSize = sizeof(bootstrap_thunk_chunk) + sizeof(bootstrap_thunk) * numThunks;
+
+    void* pbChunk = HeapAlloc(g_heapHandle, 0, chunkSize);
+
+    if (pbChunk == nullptr)
+    {
+        return false;
+    }
+
+    bootstrap_thunk_chunk* chunk = new (pbChunk) bootstrap_thunk_chunk(numThunks, (pal::dll_t)pe.GetBase());
+
+    {
+        std::lock_guard<std::mutex> _(g_thunkChunkLock);
+        chunk->SetNext(g_pVtableBootstrapThunkChunkList);
+        g_pVtableBootstrapThunkChunkList = chunk;
+    }
+
+    trace::setup();
+
+    size_t currentThunk = 0;
+    for(size_t i = 0; i < numFixupRecords; ++i)
+    {
+#ifndef _WIN64
+        assert((pFixupTable[i].Type & (COR_VTABLE_FROM_UNMANAGED | COR_VTABLE_FROM_UNMANAGED_RETAIN_APPDOMAIN)) && "managed->managed vtablefixup slots are not supported!");
```

Comment:
Why is this related to x86?

Reply:
I pulled this assert from the netfx implementation. My thought would be that it probably has to deal with something related to calling conventions. However, I don't see anything x86-specific in the runtime's fixup code, so I don't see why this assert is needed. I'll remove it.